Use `objc2-io-kit`

madsmtm opened the pull request:
Use [`objc2-io-kit`](https://docs.rs/objc2-io-kit/) and [`objc2-core-foundation`](https://docs.rs/objc2-core-foundation/) instead of [`io-kit-sys`](https://github.com/jtakakura/io-kit-rs) and [`core-foundation`](https://docs.rs/core-foundation/).

These are part of [the `objc2` project](https://github.com/madsmtm/objc2) which I maintain, which provides auto-generated bindings for Apple's APIs. This allows us to be much more confident that the bindings are correct, handles memory management like `CFRetain`/`CFRelease` automatically, allows us to do cool things like embedding documentation from the headers.

Note also that `core-foundation` has been soft-deprecated in favour of `objc2-core-foundation`, see https://github.com/servo/core-foundation-rs/issues/729, so it makes sense to start this migration.

This builds upon https://github.com/serialport/serialport-rs/pull/260, so that should probably be merged first. This change also requires an MSRV bump to 1.71 because these crates use `extern "C-unwind"`, so this should probably be put in the `v5.0.0` milestone IIUC, see also https://github.com/serialport/serialport-rs/issues/231.

madsmtm committed:
Allow compiling on tvOS, watchOS and visionOS too

Assuming that the `mach2` crate is updated to support these platforms,
or we get rid of that dependency.

madsmtm committed:
Use objc2-core-foundation and objc2-io-kit crates

madsmtm committed:
Avoid mach2 direct dependency

eldruin reviewed (commented):
Thank you for this work and your work on the `objc2` crates!
This sounds good to me.

FYI:
Yocto kirkstone (Rust 1.59.0) goes into EOL on April 2026.
The next supported release is Scathgap, which will have Rust 1.75.0.

sirhcel commented:
Hello @madsmtm, thank you very much for preparing this PR! I would like to integrate it into [playground-5.0](https://github.com/serialport/serialport-rs/tree/playground-5.0) for working towards our next major release. I did not spot an option for changing the target branch of this PR. But I bet you can. Could you please change it?

I would also rebase the changes onto the current playground-5.0 branch and as I have done or followed the edits there, you can find my conflict resolution in [sirhcel@test/objc2](https://github.com/sirhcel/serialport-rs/tree/test/objc2).

madsmtm commented:
I think I've done so now

sirhcel committed:
Put Apple dependencies in lexical order

sirhcel commented:
Thank you very much @madsmtm! I rebased the changes on the current playground and started with some nitpicking. I found out about the Matrix space from https://github.com/madsmtm/objc2#contact and will ask some more general questions over there.

sirhcel committed:
Use plane instead of class in IORegistryEntryGetParent

Both provide the same name but IORegistryEntryGetParent asks about a
plane and not a serivce class.

sirhcel committed:
Let type inference deduce the type of the `class_name` array

sirhcel committed:
Add missing check and safety comments in get_parent_device_by_type

The check for the result of calling `IOObjectGetClass` was missing. Due
due `IOObjectGetClass` presumably only writing the output buffer
partially, it should be zeroed for not exposing potentially undefined
data through `assume_init`.

sirhcel committed:
Use checked integer conversion in get_parent_device_by_type

We have them at our hands so let's use them.

sirhcel committed:
Add missing safety comment to get_string_property

sirhcel committed:
Limit scope of allowing deprecated items in available_ports

sirhcel commented:
Thank you very much for answering all my loosely related questions in the chat! I feel i'm done with my nitpicking. Also addressing `available_ports` with respect to the existing use of unsafe in this pr seems to much for the moment to me and I leave this to #356.

Could you please give my nitpicking and cleanup a review @madsmtm? I feel ready to integrate this PR and we should leave the points discussed in the chat for another round.

madsmtm commented on src/posix/enumerate.rs:
```diff
@@ -274,23 +272,36 @@ fn get_parent_device_by_type(
     loop {
-        let mut class_name = MaybeUninit::<[c_char; 128]>::uninit();
-        unsafe { IOObjectGetClass(device, class_name.as_mut_ptr()) };
+        let mut class_name = MaybeUninit::zeroed();
```
I approve of using `zeroed` over `uninit`, there's very little documentation in Apple's APIs that they don't read stuff.

madsmtm replied:
Though if you're zero-initializing, you could make it even safer by just using an array:
```rust
let mut class_name = [0; 128];
if KERN_SUCCESS != unsafe { IOObjectGetClass(device, &mut class_name) } {
    return None;
}
let name = CStr::from_bytes_until_nul(&class_name).unwrap();
```

sirhcel replied:
I looked into using an array but the types do not play out for using [`CStr::from_bytes_until_nul`](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_bytes_until_nul): This method expects `&[u8]` where dereferencing the array `&[i8]` aka `&[c_char]`. I find this interesting as [`CStr::from_ptr`](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_ptr) takes `*const i8`. This looks like we're facing some "impedance mismatch" between IO-Kit and `CStr`.

At a first glance I would prefer sticking to the current implementation in favor of getting around casting pointers with `as`. Do you see a better route from `[c_char; 128]` to `CStr`?

madsmtm replied:
Ah, you're right

> Do you see a better route from `[c_char; 128]` to `CStr`?

Not really, no.

madsmtm replied:
Wait, actually yes: You can do:
```rust
let name = class_name.map(|c| c as u8);
let name = CStr::from_bytes_until_nul(&class_name).unwrap();
```

madsmtm commented on src/posix/enumerate.rs:
```diff
@@ -274,23 +272,36 @@ fn get_parent_device_by_type(
-        let mut class_name = MaybeUninit::<[c_char; 128]>::uninit();
-        unsafe { IOObjectGetClass(device, class_name.as_mut_ptr()) };
+        let mut class_name = MaybeUninit::zeroed();
+        // SAFETY: `device` is supposed to be valid `io_object_t` and `class_name` points to a
+        // zero-initialzed buffer of the required size.
```
From my understanding, it's not a safety requirement that the `device` (and `io_object_t`s in general) is valid? It's an ID, not a pointer, and if I call `IOObjectGetClass(42, &mut class_name)`, the validity of the device will be checked by the kernel and I'll just get an error, not UB.

madsmtm replied:
(If you agree, then I'll probably be able to mark a lot of IOKit APIs as safe).

sirhcel replied:
I see the point you are making with `io_object_t` being an ID/handle. I have to admit that I have not much practical experience with using Apples APIs beyond serialport-rs. And ever after researching documentation from Apple on that, I did not find any statement on that issue. So i don't feel confident to deem this use case safe. Do you have any references on that for me?

madsmtm replied:
Not really any other reference than "`IOObjectRetain(42)` does not crash, while `CFRetain(42 as *const void)` definitely does" and "this is what I'd expect from kernel-backed objects (similar to file handles)".

(Also the only reason why those retain/release functions would return a status code).

(Actually, looking at it, it seems I've already marked these things safe in https://github.com/madsmtm/objc2/commit/0145877b8a5058a92b4e32fdbdeb68c43bf02362, and I think I stand by that reasoning).

madsmtm commented on src/posix/enumerate.rs:
```diff
@@ -402,7 +419,7 @@ cfg_if! {
                 let mut matching_services = MaybeUninit::uninit();
                 kern_result = IOServiceGetMatchingServices(
-                    kIOMasterPortDefault,
+                    kIOMainPortDefault,
```
This will raise your deployment target. I'd recommend that you keep using `kIOMasterPortDefault` and `IOMasterPort`, and just keep using `#[allow(deprecated)]`.

See also https://github.com/madsmtm/objc2/issues/824 and https://github.com/GuillaumeGomez/sysinfo/pull/1641.

sirhcel replied:
Thank you! I was not aware of this. I will leave it as-is then.
